Desktop: Allow for custom Joplin theme and Ace editor styles

devonzuegel opened the pull request:
_Discussed in https://discourse.joplinapp.org/t/custom-stylessheets/384/12?u=devonzuegel_

This allows users to define custom Ace editor styles in `editor.css`, similar to how they can define custom styles for the rendered Markdown pane in `userstyle.css`.

For example, I've added this to my `~/.config/joplindev-desktop/editor.css`:

![image](https://user-images.githubusercontent.com/6979755/69004281-ab13c300-08c5-11ea-86db-ef8ec5fc15b7.png)

... which in this case italicizes and bolds the `emphasis` tokens in the Ace editor and puts an ugly orange background in place of the default blue sidebar color:

![image](https://user-images.githubusercontent.com/6979755/69004398-6db03500-08c7-11ea-895a-48094ed185b8.png)

I'd be curious to get @laurent22's perspective though before I go further with this. Three things I'd want to improve if you are interested in this general concept before this is a real PR ready for review:
1. [FIXED] ~The naming I went with is not very elegant alongside the `userstyle.css` file. If I were starting from scratch I'd make them something like `joplin-theme.css` and `rendered-markdown.css` to make it more clear to users which is which. It might be too late to do that though since `userstyle.css` has been in use for a while, and there may be an even better option.~
2. ~It would be ideal if users could update both css files from inside the Preferences tabs, rather than having to dig around `~/.config/joplin-desktop/`. Ideally this would be direct editing from inside there for minimum navigation required, but an incremental step towards this would be to add a link from in there that when clicked opens up your default text editor to that file + a quick description of what adding to each file does.~
 
   Done! Here's how it looks: 
   <img width="50%" src="https://user-images.githubusercontent.com/6979755/69761351-fb87ed80-111b-11ea-830f-344f3c2ea11c.png" />

3. [FIXED] ~Right now, [`await this.loadCustomAceStyles`](https://github.com/laurent22/joplin/pull/2099/files#diff-00b0e649e4e574cdf55197c857ccadefR236) is called on every ACTION, rather than handled properly by the reducer. This means the styles are reloaded on every action, which is not good. I'll fix this of course.~

laurent22 commented:
Thanks @devonzuegel, I think that would be a useful change and I'd be glad to merge it when it's ready.

To load the CSS file, I think a good place would be the `start()` method of BaseApplication as that way we're sure it's only loaded once. Later if we want something more advanced, we can watch the CSS file for changes and reload it as needed.

Naming is a bit tricky indeed since userstyle.css is already taken. To be more consistent, something like Firefox would be good: They have userContent.css for the page, and userChrome.css for the UI. For now, maybe let's call the new CSS file "userchrome.css" since it's more or less consistent with "userstyle.css"?

devonzuegel commented on ReactNativeClient/lib/layout-utils.js:
```diff
@@ -1,9 +0,0 @@
-	return preferred;
-};
-
-module.exports = layoutUtils;
```
I believe this file isn't used anywhere in the codebase, so I deleted it. Lmk if I missed something with my global search though!

devonzuegel commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -1,3 +1,15 @@
+// ***                                                               ****
+// **********************************************************************
+// *********************************************************************
+
```
Discussion: https://discourse.joplinapp.org/t/re-generated-electronclient-app-lib-models-setting-js-on-build/4429/1

laurent22 replied:
I missed that discussion earlier, I've added a comment over there.

laurent22 replied:
As mentioned in the discussion over there, adding the comment won't help since this issue applies to all the files, not just this one (and I prefer not to add a header to all the files). For now I would suggest we look into this issue separately from this pull request.

devonzuegel replied:
Apologies, I did miss this! Thanks for [removing](https://github.com/laurent22/joplin/pull/2099#issuecomment-565251144).

laurent22 commented:
Thanks for the update @devonzuegel! I didn't go through all the diffs yet but will do soon.

tessus commented:
I like the new change, but I'm a bit puzzled by the naming:

- Custom stylesheet for rendered markdown -> userchrome.css
- Custom stylesheet for Joplin-wide app styles -> userstyle.css

Until now, we only used `userstyle.css` which basically allowed us to change how the rendered markdown will look like. But it now looks like the `userstyle.css` is for something else and `userchrome.css` is supposed to be used to change the rendered markdown.

Maybe I'm missing something.

devonzuegel commented:
> I like the new change, but I'm a bit puzzled by the naming:
> 
> * Custom stylesheet for rendered markdown -> userchrome.css
> * Custom stylesheet for Joplin-wide app styles -> userstyle.css

You're right, I accidentally swapped the names! Fixed. Thank you for catching, that would've been super confusing to users.



> Update: I think it should rather be ike this, unless I completely misunderstood the reason for this PR:
> 
> * Custom stylesheet for markdown tokens (editor) -> userchrome.css
> * Custom stylesheet for rendered markdown (preview) -> userstyle.css

The reason for the PR was partially to support markdown tokens, but the purpose is broader than just that—the Joplin-wide styles support customization of the whole app via CSS file, including but not limited to the markdown tokens. So for example, you could change the sidebar color with this if you'd like too, which is lighter weight than creating a full theme change.

tessus commented on ElectronClient/app/gui/ConfigScreen.jsx:
```diff
@@ -413,6 +413,24 @@ class ConfigScreenComponent extends React.Component {
+						<label>{md.label()}</label>
+					</div>
+					<button style={buttonStyle} onClick={md.onClick}>
+						Edit
```
This should be a translatable string.

devonzuegel replied:
Oh you are right. Besides calling the `_(...)` function, how do I go about adding that translation?

tessus replied:
You don't have to do anything other than wrapping it in the `_(` function. The string `Edit` should already have been translated and added to the .pot file. After the PR, we have to update the translations anyway, because you added new strings as well.

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -410,6 +423,37 @@ class Setting extends BaseModel {
+				type: Setting.TYPE_BUTTON,
+				public: true,
+				appTypes: ['desktop'],
+				label: () => _('Custom stylesheet for rendered markdown'),
```
For consistency, "Markdown" should be with a capital letter.

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -410,6 +423,37 @@ class Setting extends BaseModel {
+			// a way to display a link to the customizable stylesheets, not for it to
+			// serve as a customizable Setting. But because the Setting page is auto-
+			// generated from this list of settings, there wasn't a really elegant way
+			// to do that directly in the React markup.
```
I think that's good actually, and the "button" setting type could be useful for other things too. Thanks for implementing this.

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -8,6 +20,7 @@ const { toTitleCase } = require('lib/string-utils.js');
 const { rtrimSlashes } = require('lib/path-utils.js');
 const { _, supportedLocalesToLanguages, defaultLocale } = require('lib/locale.js');
 const { shim } = require('lib/shim');
+const { bridge } = require('electron').remote.require('./bridge');
```
Oh sorry I missed that. We can't have a dependency to the Electron package here as it needs to compile in React Native too. I think it should work if you use `shim.openUrl` instead.

devonzuegel replied:
Good catch! Thanks.

devonzuegel commented:
All righty @laurent22, I incorporated your comments. I think this is ready for final review. Thank you!

eschafer commented:
I'm really excited for this to go through - thanks for working on it!

The only thing I noticed was that nothing happens when you click on the preference buttons if you don't already have style sheets in the config folder. It'd be nice if those files could be created automatically, but if not, some sort of messaging might help.

Actually one more thing that might be out of scope: it'd be nice if the dev tools showed the contents of the whole site rather than just the note preview. Without that, I have to build the app myself or look through github to figure out which selectors I can use.

devonzuegel commented:
> It'd be nice if those files could be created automatically, but if not, some sort of messaging might help.

Great idea @eschafer! Updated the PR so that now it creates the file before opening it if it's not there.

> Actually one more thing that might be out of scope: it'd be nice if the dev tools showed the contents of the whole site rather than just the note preview. Without that, I have to build the app myself or look through github to figure out which selectors I can use.

If you run the [local build](https://github.com/laurent22/joplin/blob/master/BUILD.md#building-the-electron-application) you can get two Developer Tools windows, one for the rendered markdown (like in the normal pre-built app, which is styled by `userstyle.css`) and one for the rest of the app (styled by `userchrome.css`, which isn't usually available in the pre-built app for some reason). 

Here's how it looks (right = rendered Markdown dev tools, bottom = dev tools for the rest of the app):

![image](https://user-images.githubusercontent.com/6979755/70499852-eec9a900-1ace-11ea-9c33-90f707b6454e.png)

---

One issue with the `userstyle.css` and `userchrome.css` naming is that by default they open with Firefox (or when I removed Firefox, Safari). This isn't great because they're not editable directly from there. I'd prefer they open in my text editor, so there's no additional step to open them somewhere editable. Any thoughts on this @laurent22? I don't think this is blocking for this PR, but not optimal.

laurent22 commented:
> One issue with the userstyle.css and userchrome.css naming is that by default they open with Firefox (or when I removed Firefox, Safari). This isn't great because they're not editable directly from there. I'd prefer they open in my text editor, so there's no additional step to open them somewhere editable. Any thoughts on this @laurent22? I don't think this is blocking for this PR, but not optimal.

I think this is fine for now but indeed it could be improved. Perhaps we could use the configured external editor to open these CSS files?

Another issue was the header on Setting.js - I've left [a comment](https://github.com/laurent22/joplin/pull/2099#discussion_r353120915) which you might have missed, so for now I've removed the header, and we can maybe look at improving the situation with the lib folder in a different PR.

Otherwise it's all good and ready to merge!

nemesis5 commented:
Hi  devonzuegel, hi laurent,

thank you for your great work here. This feature will help me to correct some theming issues localy, for example inside the new nord theme.

The only issue I have is that I cannot see which css parts are available, because in the mentioned pre built app the dev tools for the "rest of the app" are not availalbe.

My question is why you hide this dev tools (the official dev tools are only shown for the rendered markdown part)?
Couldn't you show the dev tools for the rest of the app as they are enabled inside the tools menu?

This would help everybody to get the most out of this custom css feature implemented by devonzuegel.

Thank you.

devonzuegel commented:
Hi @nemesis5. Thanks for the nice words! I'm excited about this feature too.

If you follow the directions in the [Debugging the Desktop Application](https://joplinapp.org/debugging/#desktop-application) docs, you will be able to see the dev tools. @laurent22 what do you think of making this more directly accessible via a menu?

laurent22 commented:
@devonzuegel, in the next version (already released as a pre-release), the notes will be rendered in an iframe so there will be just a single dev tool for the app and note viewer. The change is already made and the tool will be under the Help menu.

devonzuegel commented:
Oh lovely, that's an improvement! Thanks for letting me know.

devonzuegel commented:
I've created a little repo of my custom styles in case they're helpful to anyone else:
https://discourse.joplinapp.org/t/collection-of-custom-css/5428
